### PR TITLE
Add support for linear in mkldnn fusion

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52614 Avoid dispatch overhead in call to mkldnn convolution
* #52613 Include max pool in fusion groups
* #52513 Extend subgraph utils to cover merging a node following a subgraph
* #52512 Remove unused subgraph vmap api
* #51924 Only functionalize ops which we want to include in mkldnn group
* #51923 Add a filter to remove mutation
* #51922 Dont error on 0-dim in convolution
* #51921 Add 0-dim support for binary mkldnn ops
* #51736 Handle mkldnn broadcasting in mkldnn fuser
* #51600 Add MKLDNN fuser
* #52786 [NNC] Make NNC sanitize input names
* **#51484 Add support for linear in mkldnn fusion**
* #51483 Move conv to mkldnn

This PR moves the linear weights of a frozen model to MKLDNN. When the weights are already in MKLDNN, just computing a single linear by converting the input and output from/to mkldnn provides large speedups. I benchmark'd the results of the top 200 shapes in predictor [here](https://www.internalfb.com/phabricator/paste/view/P171537854) (taken from aten::matmul), as well as verified that it sped up popular models. .

Differential Revision: [D26696698](https://our.internmc.facebook.com/intern/diff/D26696698)

### CODE REVIEW

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp [hunk 1]
```diff
@@ -138,66 +162,89 @@ bool frozenMkldnnCompatibleConvNode(Node* n) {
...
+  n->replaceInputWith(input, mkldnn_tensor_value);
+}
+
+void replaceInputWithMKLDNNTensor(Node* n, std::string name) {
+  Value* input = n->namedInput(name);
+  auto mkldnn_tensor = constant_as<Tensor>(input)->to_mkldnn();
+  replaceInputWithMKLDNNTensor(n, name, mkldnn_tensor);
+}
+
+void convertFrozenConvToMKLDNN(Node* conv) {
```
Comment (Krovatkin) [inline torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:187]: did anything change in here from the previous PR?

File: test/jit/test_freezing.py [hunk 2]
```diff
@@ -1563,6 +1563,48 @@ def to_dense(inp):
...
 
+    @unittest.skipIf(not torch._C.has_mkldnn, "MKL-DNN build is disabled")
+    def test_linear_to_mkldnn(self):
+        # use until github.com/pytorch/pytorch/pull/50856 land
+        class LinearMod(nn.Linear):
+            def __init__(self, *args, **kwargs):
+                super().__init__(*args, **kwargs)
+
+            def forward(self, input):
+                return torch._C._nn.linear(input, self.weight, self.bias)
```
Comment (Krovatkin) [inline test/jit/test_freezing.py:1574]: this is to get rid of an if?

File: aten/src/ATen/native/mkldnn/Linear.cpp [hunk 3]
```diff
@@ -28,16 +28,22 @@ Tensor mkldnn_linear(
     const Tensor& self,
     const Tensor& weight,
     const Tensor& bias) {
-  TORCH_CHECK(self.dim() >= 2,
-      "mkldnn_linear: input needs to has dim at least 2, input dim ", self.dim());
+  const int64_t dim = self.dim();
+  TORCH_CHECK(
+      self.dim() != 0,
```
Comment (Krovatkin) [inline aten/src/ATen/native/mkldnn/Linear.cpp:33]: I wonder if there was a performance concern here. 
MKLDNN also has `inner_product`: https://oneapi-src.github.io/oneDNN/structmkldnn_1_1inner__product__forward_1_1desc.html#details We could ask Intel guys.

**Review by Krovatkin — APPROVED** [review]
:shipit:

File: aten/src/ATen/native/mkldnn/Linear.cpp (hunk 3 above)
Comment (eellison) [reply aten/src/ATen/native/mkldnn/Linear.cpp:33]: Yea, that's a good idea. With dim > 2 they just do a reshape, so this is equivalent to that approach. From looking at the top 200 shapes internally broadcasting linear is extremely rare. I think `inner_product` might be a good follow up but not necessary for this PR..

File: test/jit/test_freezing.py (hunk 2 above)
Comment (eellison) [reply test/jit/test_freezing.py:1574]: This is to compile it as one op. This is how it's done on master as of now but wasnt landed yet when i submitted this PR.

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp (hunk 1 above)
Comment (eellison) [reply torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:187]: No, just refactoring..
